### PR TITLE
feat(server): dont use channel for replication / save df

### LINKED ISSUE #4040: Pass io::Sink to snapshot in case we perform DFS/Replication snapshotting, and give up on the channel.

### PR DESCRIPTION
**Pull request by adiholden**
fixes #4040

### CODE REVIEW

File: src/server/rdb_save.h [hunk 1]
```diff
@@ -89,21 +89,26 @@ class RdbSaver {
...
+  std::error_code StopFullSyncInShard(EngineShard* shard);
+
+  // Wait for snapshoting finishe in shard thread. Called from save flows in shard thread.
+  std::error_code WaitSnapshotInShard(EngineShard* shard);
 
   // Stores auxiliary (meta) values and header_info
   std::error_code SaveHeader(const GlobalData& header_info);
 
   // Writes the RDB file into sink. Waits for the serialization to finish.
+  // Called only for save rdb flow and save df on summery file.
```
Comment (romange) [inline src/server/rdb_save.h:107]: summery -> summary

File: src/server/rdb_save.h [hunk 2]
```diff
@@ -89,21 +89,26 @@ class RdbSaver {
...
+  void StartSnapshotInShard(bool stream_journal, Context* cntx, EngineShard* shard);
 
   // Send only the incremental snapshot since start_lsn.
   void StartIncrementalSnapshotInShard(Context* cntx, EngineShard* shard, LSN start_lsn);
 
   // Stops full-sync serialization for replication in the shard's thread.
-  void StopFullSyncInShard(EngineShard* shard);
+  std::error_code StopFullSyncInShard(EngineShard* shard);
+
+  // Wait for snapshoting finishe in shard thread. Called from save flows in shard thread.
```
Comment (romange) [inline src/server/rdb_save.h:100]: `finishe`

File: src/server/rdb_save.cc [hunk 3]
```diff
@@ -1213,39 +1232,38 @@ error_code RdbSaver::Impl::SaveAuxFieldStrStr(string_view key, string_view val)
 
 error_code RdbSaver::Impl::ConsumeChannel(const Cancellation* cll) {
   error_code io_error;
-  SliceSnapshot::DbRecord record;
```
Comment (romange) [inline src/server/rdb_save.cc:1216]: maybe add a comment stating it's been used only for legacy rdb save flows.

File: src/server/rdb_save.cc [hunk 4]
```diff
@@ -1278,26 +1296,57 @@ error_code RdbSaver::Impl::WriteRecord(io::Bytes src) {
   return ec;
 }
 
-void RdbSaver::Impl::StartSnapshotting(bool stream_journal, const Cancellation* cll,
-                                       EngineShard* shard) {
+void RdbSaver::Impl::PushSnapshotData(Context* cntx, string record) {
+  if (cntx->IsCancelled()) {
+    return;
+  }
+  if (save_mode_ == SaveMode::RDB) {  // Rdb write to channle
```
Comment (romange) [inline src/server/rdb_save.cc:1303]: `channel`

Comment (romange) [inline src/server/rdb_save.cc:1303]: you use `save_mode_ == SaveMode::RDB` when you create the channel. here you can just check `if (channel_)` to localize business logic into a single place in the code.

File: src/server/rdb_save.cc [hunk 5]
```diff
@@ -1278,26 +1296,57 @@ error_code RdbSaver::Impl::WriteRecord(io::Bytes src) {
...
+    auto ec = WriteRecord(io::Buffer(record));
+    if (ec) {
+      cntx->ReportError(ec);
+    }
+  }
+}
+
+void RdbSaver::Impl::FinalizeSnapshotWriting() {
+  if (save_mode_ == SaveMode::RDB) {
+    channel_->StartClosing();
```
Comment (romange) [inline src/server/rdb_save.cc:1315]: same here.

File: src/server/rdb_save.cc [hunk 6]
```diff
@@ -1278,26 +1296,57 @@ error_code RdbSaver::Impl::WriteRecord(io::Bytes src) {
...
+
+  s->Start(stream_journal, cntx->GetCancellation(), allow_flush);
 }
 
 void RdbSaver::Impl::StartIncrementalSnapshotting(Context* cntx, EngineShard* shard,
                                                   LSN start_lsn) {
   auto& db_slice = namespaces.GetDefaultNamespace().GetDbSlice(shard->shard_id());
   auto& s = GetSnapshot(shard);
-  s = std::make_unique<SliceSnapshot>(&db_slice, &channel_, compression_mode_);
+  auto on_finalize_cb = std::bind(&RdbSaver::Impl::FinalizeSnapshotWriting, this);
```
Comment (romange) [inline src/server/rdb_save.cc:1337]: it's possible to do like this. not critical but when you pass two lambdas, maybe it's a sign that using an virtual interface is a better pattern here, and make `RdbSaver::Impl` derive from it.

File: src/server/snapshot.h [hunk 7]
```diff
@@ -31,35 +30,27 @@ struct Entry;
...
 //              │
-//              |                      Channel is left open in journal streaming mode
+//              |                      Socket is left open in journal streaming mode
 //              ▼
 // ┌──────────────────────────┐          ┌─────────────────────────┐
 // │     SerializeEntry       │ ◄────────┤     OnJournalEntry      │
 // └─────────────┬────────────┘          └─────────────────────────┘
 //               │
-//         PushBytesToChannel        Default buffer gets flushed on iteration,
+//         PushBytes        Default buffer gets flushed on iteration,
```
Comment (romange) [inline src/server/snapshot.h:39]: indentation

File: src/server/rdb_save.cc [hunk 8]
```diff
@@ -1344,16 +1344,23 @@ void RdbSaver::Impl::StartIncrementalSnapshotting(Context* cntx, EngineShard* sh
 
 // called on save flow
 void RdbSaver::Impl::WaitForSnapshottingFinish(EngineShard* shard) {
-  GetSnapshot(shard)->WaitSnapshoting();
+  auto& snapthot = GetSnapshot(shard);
+  CHECK(snapthot);
+  snapthot->WaitSnapshoting();
```
Comment (romange) [inline src/server/rdb_save.cc:1349]:
```suggestion
  auto& snapshot = GetSnapshot(shard);
  CHECK(snapshot);
  snapshot->WaitSnapshoting();
```

File: src/server/rdb_save.cc [hunk 9]
```diff
@@ -1344,16 +1344,23 @@ void RdbSaver::Impl::StartIncrementalSnapshotting(Context* cntx, EngineShard* sh
...
 }
 
 // called from replication flow
 void RdbSaver::Impl::StopSnapshotting(EngineShard* shard) {
-  GetSnapshot(shard)->FinalizeJournalStream(false);
+  auto& snapthot = GetSnapshot(shard);
+  CHECK(snapthot);
+  snapthot->FinalizeJournalStream(false);
```
Comment (romange) [inline src/server/rdb_save.cc:1356]:
```suggestion
  auto& snapshot = GetSnapshot(shard);
  CHECK(snapshot);
  snapshot->FinalizeJournalStream(false);
```

File: src/server/rdb_save.cc [hunk 10]
```diff
@@ -1344,16 +1344,23 @@ void RdbSaver::Impl::StartIncrementalSnapshotting(Context* cntx, EngineShard* sh
...
 }
 
 void RdbSaver::Impl::CancelInShard(EngineShard* shard) {
-  GetSnapshot(shard)->FinalizeJournalStream(true);
+  auto& snapthot = GetSnapshot(shard);
+  if (snapthot) {  // Cancel can be called before snapshoting started.
+    snapthot->FinalizeJournalStream(true);
```
Comment (romange) [inline src/server/rdb_save.cc:1362]:
```suggestion
  auto& snapshot = GetSnapshot(shard);
  if (snapshot) {  // Cancel can be called before snapshotting started.
    snapshot->FinalizeJournalStream(true);
```

File: src/server/detail/save_stages_controller.cc [hunk 11]
```diff
@@ -395,13 +408,22 @@ GenericError SaveStagesController::BuildFullPath() {
   return {};
 }
 
-void SaveStagesController::SaveCb(unsigned index) {
-  if (auto& snapshot = snapshots_[index].first; snapshot && snapshot->HasStarted())
+void SaveStagesController::SaveBody(unsigned index) {
+  CHECK(!use_dfs_format_ || index == shard_set->size());
```
Comment (romange) [inline src/server/detail/save_stages_controller.cc:412]: this CHECK is weird `index == shard_set->size()`

File: src/server/snapshot.h [hunk 12]
```diff
@@ -85,7 +76,7 @@ class SliceSnapshot {
 
   // Waits for a regular, non journal snapshot to finish.
   // Called only for non-replication, backups usecases.
-  void Join() {
+  void WaitSnapshoting() {
```
Comment (romange) [inline src/server/snapshot.h:79]: WaitSnapshoting -> `WaitSnapshotting`
please check if you have more `shoting`  cases

**Review by romange — COMMENTED** [review]
LGTM

I think it will be more readable if we deduplicate some of the RdbSerializer functions so that we won't have two sets of functions - one for replication and one for snapshotting.

File: src/server/snapshot.h (hunk 12 above)
Comment (adiholden) [reply src/server/snapshot.h:79]: done

**Comment by adiholden** [comment]
> LGTM
> 
> I think it will be more readable if we deduplicate some of the RdbSerializer functions so that we won't have two sets of functions - one for replication and one for snapshotting.

I will do addittional refactoring changes in a separate PR ok?

**Comment by romange** [comment]
Yeah sure, it was a general comment 👌

**Review by romange — DISMISSED** [review]

**Review by romange — APPROVED** [review]

**Pull request merged**